File: sublimejava.py
"""
Copyright (c) 2012 Fredrik Ehnbom

This software is provided 'as-is', without any express or implied
warranty. In no event will the authors be held liable for any damages
arising from the use of this software.

Permission is granted to anyone to use this software for any purpose,
including commercial applications, and to alter it and redistribute it
freely, subject to the following restrictions:

   1. The origin of this software must not be misrepresented; you must not
   claim that you wrote the original software. If you use this software
   in a product, an acknowledgment in the product documentation would be
   appreciated but is not required.

   2. Altered source versions must be plainly marked as such, and must not be
   misrepresented as being the original software.

   3. This notice may not be removed or altered from any source
   distribution.
"""
import sublime
import sublime_plugin
import os
import re
import bisect
import imp

def reload(mod):
    n = mod.__file__
    if n[-1] == 'c':
        n = n[:-1]
    globals()[mod.__name__] = imp.load_source(mod.__name__, n)
completioncommon = imp.load_source("completioncommon", os.path.join(os.path.dirname(os.path.abspath(__file__)), "sublimecompletioncommon/completioncommon.py"))
completioncommon.reload(completioncommon)

try:
    import SublimeJava.classopener as classopener
except:
    import classopener
imp.reload(classopener)


class SublimeJavaDotComplete(completioncommon.CompletionCommonDotComplete):
    pass


class SublimeJavaCompletion(completioncommon.CompletionCommon):
    def __init__(self):
        super(SublimeJavaCompletion, self).__init__("SublimeJava.sublime-settings", os.path.dirname(os.path.abspath(__file__)))
        self.regex = [
            (re.compile(r"\[I([,)}]|$)"), r"int[]\1"),
            (re.compile(r"\[F([,)}]|$)"), r"float[]\1"),
            (re.compile(r"\[Z([,)}]|$)"), r"boolean[]\1"),
            (re.compile(r"\[B([,)}]|$)"), r"byte[]\1"),
            (re.compile(r"\[C([,)}]|$)"), r"char[]\1"),
            (re.compile(r"\[S([,)}]|$)"), r"short[]\1"),
            (re.compile(r"\[J([,)}]|$)"), r"long[]\1"),
            (re.compile(r"\[D([,)}]|$)"), r"double[]\1"),
            (re.compile(r"\[\L?([\w\./]+)(;)?"), r"\1[]")]

    def show_error(self, msg):
        if self.get_setting("sublimejava_no_visual_errors", False):
            print(msg)
        else:
            sublime.error_message(msg + "\n\nDisable visual error message dialogues with setting:\nsublimejava_no_visual_errors: true")

    def get_packages(self, data, thispackage, type):
        packages = re.findall(r"(?:^|\n)[ \t]*import[ \t]+(.*);", data)
        packages.append("java.lang.*")
        packages.append("")  # for int, boolean, etc
        for package in packages:
            idx = type.find(".")
            if idx == -1:
                idx = len(type)
            subtype = type[:idx]
            if re.search("[\.\$]{1}%s$" % subtype, package):
                # Explicit imports, we want these to have the highest
                # priority when searching for the absolute type, so
                # insert them at the top of the package list.
                # Both the .* version and not is added so that
                # blah.<searchedForClass> and blah$<searchedForClass>
                # is tested
                add = package[:-(len(subtype)+1)]
                packages.insert(0, add + ".*")
                packages.insert(1, add)
                break
        packages.append(thispackage + ".*")
        return packages

    def get_cmd(self):
        classpath = self.get_setting("sublimejava_classpath", ["."])
        newclasspath = []
        window = sublime.active_window()
        for path in classpath:
            newclasspath.append(self.expand_path(path, window))
        classpath = newclasspath
        classpath.insert(0, ".")
        classpath = os.pathsep.join(classpath)
        return "java -classpath \"%s\" SublimeJava" % classpath

    def is_supported_language(self, view):
        if view.is_scratch() or not self.get_setting("sublimejava_enabled", True):
            return False
        language = self.get_language(view)
        return language == "java" or language == "jsp"

    def sub(self, regex, sub, data):
        olddata = data
        data = regex.sub(sub, data)
        while data != olddata:
            olddata = data
            data = regex.sub(sub, data)
        return data

    def fixnames(self, data):
        for regex, replace in self.regex:
            data = self.sub(regex, replace, data)
        return data

    def return_completions(self, comp):
        ret = []
        for display, insert in comp:
            ret.append((self.fixnames(display), self.fixnames(insert)))
        return super(SublimeJavaCompletion, self).return_completions(ret)

    def get_class_under_cursor(self):
        view = sublime.active_window().active_view()
        data = view.substr(sublime.Region(0, view.size()))
        word = view.substr(view.word(view.sel()[0].begin()))
        return self.find_absolute_of_type(data, data, word)

    def get_possible_imports(self, classname):
        imports = []

        if classname is not None:
            stdout = self.run_completion("-possibleimports;;--;;%s" % classname)
            imports = sorted(stdout.split("\n")[:-1])

        return imports


comp = SublimeJavaCompletion()


class SublimeJava(sublime_plugin.EventListener):

    def on_query_completions(self, view, prefix, locations):
        return comp.on_query_completions(view, prefix, locations)

    def on_query_context(self, view, key, operator, operand, match_all):
        if key == "sublimejava.dotcomplete":
            return comp.get_setting(key.replace(".", "_"), True)
        elif key == "sublimejava.supported_language":
            return comp.is_supported_language(view)
        else:
            return comp.on_query_context(view, key, operator, operand, match_all)


MSG_NO_CLASSES_FOUND = "No classes found to import for name %s."
MSG_ALREADY_IMPORTED = "Class %s has either already been imported, is \
in the current package, or is in the default package."

RE_IMPORT = "import( static)? ([\w\.]+)\.([\w]+|\*);"
RE_PACKAGE = "package ([\w]+.)*\w+;"
RE_IMPORT_SECTION = "(^import[^;\n]+;[^\n]*\n)+"


class ExecuteImportJavaCommand(sublime_plugin.TextCommand):
    def run(self, edit, full_classname):
        insert_point = 0
        newlines_prepend = 0
        newlines_append = 1

        import_classname = full_classname.replace("$", ".")
        import_statement = "import %s;" % (import_classname)
        all_imports_region = self.view.find_all(RE_IMPORT)

        if len(all_imports_region) > 0:
            all_imports = sorted([(region, self.view.substr(region)) for region in all_imports_region], key=lambda a: a[1])
            only_imports = [a[1] for a in all_imports]

            # The following logic is used to find the right spot to insert the import statement at if there are multiple
            # separate import groups. See discussion in #62
            pos = bisect.bisect_left(only_imports, import_statement)

            def score_string(a, b):
                score = 0
                for i in range(min(len(a), len(b))):
                    score += a[i] == b[i]
                return score

            if pos == len(all_imports) or (pos > 0 and \
                    score_string(import_statement, all_imports[pos-1][1]) > score_string(import_statement, all_imports[pos][1])):
                # Insert after
                insert_point = all_imports[pos-1][0].b
                newlines_prepend = 1
                newlines_append = 0
            else:
                # Insert before
                insert_point = all_imports[pos][0].a
                newlines_prepend = 0
                newlines_append = 1
        else:
            package_declaration_region = self.view.find(RE_PACKAGE, 0)

            if package_declaration_region is not None:
                insert_point = package_declaration_region.b
                newlines_prepend = 2
                newlines_append = 0

        import_statement = "%s%s%s" % ("\n" * newlines_prepend,
                                               import_statement,
                                               "\n" * newlines_append)

        self.view.insert(edit, insert_point, import_statement)

        if comp.get_setting("sublimejava_organize_imports", True):
            self.view.run_command("organize_java_imports")


class ImportJavaClassCommand(sublime_plugin.TextCommand):
    def run(self, edit):
        view = self.view
        classname = view.substr(view.word(view.sel()[0].begin()))

        if comp.get_class_under_cursor():
            comp.show_error(MSG_ALREADY_IMPORTED % classname)
            return

        imports = comp.get_possible_imports(classname)

        def do_import(index):
            if index != -1:
                self.view.run_command("execute_import_java", {"full_classname": imports[index]})

        if len(imports) > 0:
            view.window().show_quick_panel(imports, do_import)
        else:
            comp.show_error(MSG_NO_CLASSES_FOUND % classname)


class OrganizeJavaImportsCommand(sublime_plugin.TextCommand):
    def run(self, edit):
        sections = self.view.find_all(RE_IMPORT_SECTION, 0)
        section_imports = [self.view.substr(section) for section in sections]
        for i in range(len(sections)):
            # TODO pass func to organize as java, 3rd party, project
            imports = section_imports[i][:-1].split("\n")
            imports.sort()
            imports = "\n".join(imports) + "\n"

            self.view.replace(edit, sections[i], imports)


class OpenJavaSourceCommand(sublime_plugin.WindowCommand):

    def run(self, under_cursor=False):
        classopener.JavaSourceOpener(comp,
                                     self.window.active_view(),
                                     under_cursor).show()


class OpenJavaDocCommand(sublime_plugin.WindowCommand):

    def run(self, under_cursor=False):
        classopener.JavaDocOpener(comp,
                                  self.window.active_view(),
                                  under_cursor).show()

File: SublimeJava.java
/*
Copyright (c) 2012 Fredrik Ehnbom

This software is provided 'as-is', without any express or implied
warranty. In no event will the authors be held liable for any damages
arising from the use of this software.

Permission is granted to anyone to use this software for any purpose,
including commercial applications, and to alter it and redistribute it
freely, subject to the following restrictions:

   1. The origin of this software must not be misrepresented; you must not
   claim that you wrote the original software. If you use this software
   in a product, an acknowledgment in the product documentation would be
   appreciated but is not required.

   2. Altered source versions must be plainly marked as such, and must not be
   misrepresented as being the original software.

   3. This notice may not be removed or altered from any source
   distribution.
*/
import java.lang.reflect.*;
import java.net.*;
import java.io.*;
import java.util.*;
import java.util.regex.Pattern;
import java.util.regex.Matcher;
import java.util.jar.*;

public class SublimeJava
{

    private static String getInstancedType(Class<?> c, String gen, String ret, String[] templateParam)
    {
        if (gen.startsWith("class "))
            gen = gen.substring("class ".length());
        {
            // This is a bit odd, and I'm not sure it's correct. Seems "gen" returns an absolute
            // class path that isn't correct. I've only seen this for the specific class that is
            // in the repro case of issue #26, but it probably happens elsewhere too so could
            // have to be tweaked.
            //
            // In short, gen will list Tests.Tests$Foo whereas the correct type would be just
            // Tests$Foo
            String pat = "(\\w+\\.)+"+ Pattern.quote(ret);
            Pattern p = Pattern.compile(pat);
            Matcher m = p.matcher(gen);
            if (m.find())
            {
                gen = m.replaceAll(Matcher.quoteReplacement(ret));
            }
        }

        if (!gen.equals(ret))
        {
            boolean set = false;
            TypeVariable<?> tv[] = c.getTypeParameters();
            for (int i = 0; i < tv.length && i < templateParam.length; i++)
            {
                String pat = "((^|[<,\\s])+)" + tv[i].getName() +  "(([,\\s>]|$)+)";
                Pattern p = Pattern.compile(pat);
                Matcher m = p.matcher(gen);
                if (m.find())
                {
                    gen = m.replaceAll(Matcher.quoteReplacement(m.group(1) + templateParam[i] + m.group(3)));
                }
            }
            ret = gen;
        }
        return ret;
    }

    private static String[] getCompletion(Method m, String filter, String[] templateParam)
    {
        String str = m.getName();
        if (!str.startsWith(filter))
            return null;
        str += "(";
        String ins = str;
        int count = 1;
        Type[] generic = m.getGenericParameterTypes();
        Class[] normal = m.getParameterTypes();
        for (int i = 0; i < normal.length; i++)
        {
            if (count > 1)
            {
                str += ", ";
                ins += ", ";
            }

            String gen = generic[i].toString();
            String ret = normal[i].getName();
            ret = getInstancedType(m.getDeclaringClass(), gen, ret, templateParam);
            str += ret;
            ins += "${"+count + ":" + ret.replace("$", "\\$") + "}";
            count++;
        }
        str += ")\t" + getInstancedType(m.getDeclaringClass(), m.getGenericReturnType().toString(), m.getReturnType().getName(), templateParam);
        ins += ")";
        return new String[] {str, ins};
    }
    private static String[] getCompletion(Field f, String filter, String[] templateArgs)
    {
        String str = f.getName();
        if (!str.startsWith(filter))
            return null;

        String rep = str + "\t" + getInstancedType(f.getDeclaringClass(), f.getGenericType().toString(), f.getType().getName(), templateArgs);
        return new String[] {rep, str};
    }
    private static String[] getCompletion(Class clazz, String filter, String[] templateArgs)
    {
        return new String[] {clazz.getSimpleName() + "\tclass", clazz.getSimpleName()};
    }
    private static <T> String[] getCompletion(T t, String filter, String[] templateArgs)
    {
        if (t instanceof Method)
        {
            return getCompletion((Method)t, filter, templateArgs);
        }
        else if (t instanceof Field)
        {
            return getCompletion((Field)t, filter, templateArgs);
        }
        else if (t instanceof Class)
        {
            return getCompletion((Class)t, filter, templateArgs);
        }
        return null;
    }
    private static final String sep = ";;--;;";

    private static String getClassname(String pack, String clazz)
    {
        if (pack.endsWith(".*"))
        {
            return pack.substring(0, pack.length()-2) + "." + clazz;
        }
        else if (pack.length() != 0)
        {
            return pack + "$" + clazz;
        }
        return clazz;
    }

    private static final int STATIC_BIT = 1 << 0;
    private static final int PRIVATE_BIT = 1 << 1;
    private static final int PROTECTED_BIT = 1 << 2;
    private static final int PUBLIC_BIT = 1 << 3;

    private static <T> int getModifiers(T t)
    {
        int modifiers = 0;
        if (t instanceof Method)
        {
            Method m = (Method) t;
            modifiers = m.getModifiers();
        }
        else if (t instanceof Field)
        {
            Field f = (Field) t;
            modifiers = f.getModifiers();
        }
        else if (t instanceof Class)
        {
            Class c = (Class) t;
            modifiers = c.getModifiers();
        }
        int returnvalue = 0;
        if (Modifier.isStatic(modifiers))
            returnvalue |= STATIC_BIT;
        if (Modifier.isPrivate(modifiers))
            returnvalue |= PRIVATE_BIT;
        if (Modifier.isProtected(modifiers))
            returnvalue |= PROTECTED_BIT;
        if (Modifier.isPublic(modifiers))
            returnvalue |= PUBLIC_BIT;
        return returnvalue;
    }

    private static <T> void dumpCompletions(T[] arr, String filter, String[] templateArgs)
    {
        for (T t : arr)
        {
            String[] completion = getCompletion(t, filter, templateArgs);
            if (completion == null)
            {
                continue;
            }
            System.out.println(completion[0] + sep + completion[1] + sep + getModifiers(t));
        }
    }

    private static boolean getReturnType(Field[] fields, String filter, String templateParam[])
    {
        for (Field f : fields)
        {
            if (filter.equals(f.getName()))
            {
                String gen = f.getGenericType().toString();
                String ret = f.getType().getName();
                ret = getInstancedType(f.getDeclaringClass(), gen, ret, templateParam);
                System.out.println("" + ret);
                return true;
            }
        }
        return false;
    }

    private static boolean getReturnType(Method[] methods, String filter, String templateParam[])
    {
        for (Method m : methods)
        {
            if (filter.equals(m.getName()))
            {
                String gen = m.getGenericReturnType().toString();
                String ret = m.getReturnType().getName();
                ret = getInstancedType(m.getDeclaringClass(), gen, ret, templateParam);
                System.out.println("" + ret);
                return true;
            }
        }
        return false;
    }
    private static boolean getReturnType(Class[] classes, String filter, String templateParam[])
    {
        for (Class clazz : classes)
        {
            if (filter.equals(clazz.getSimpleName()))
            {
                System.out.println(clazz.getName());
                return true;
            }
        }
        return false;
    }

    private static boolean isPackage(String packageName)
        throws IOException
    {
        Package p = Package.getPackage(packageName);
        if (p != null)
            return true;

        packageName = packageName.replace(".", File.pathSeparator);

        ClassLoader classLoader = Thread.currentThread().getContextClassLoader();
        for (String s : getClasspathEntries())
        {
            URL url = classLoader.getResource(s + File.pathSeparator + packageName);
            if (url != null)
                return true;
            else
                url = classLoader.getResource(s);
            if (url == null)
                continue;

            String filename = URLDecoder.decode(url.getFile(), "UTF-8");

            if (url.getProtocol().equals("jar"))
            {
                filename = filename.substring(5, filename.indexOf("!"));

                JarFile jf = new JarFile(filename);
                Enumeration<JarEntry> entries = jf.entries();
                while (entries.hasMoreElements())
                {
                    String name = entries.nextElement().getName();
                    if (name.startsWith(packageName))
                    {
                        return true;
                    }
                }
            }
            else
            {
                File folder = new File(filename + File.pathSeparator + packageName);
                if (folder.exists())
                    return true;
            }
        }
        return false;
    }

    private static final String CLASS_FILE_NAME_RE = "(?:$|/)?(\\w+).class";
    private static final String CLASS_FULL_NAME_IN_PKG_RE = "(\\w+(\\.|$){1})+\\w+";
    private static final String DIGITS_CLASS_NAME_RE = "\\d+";
    private static final String SUBLIME_JAVA_CLASS_RE = "SublimeJava";

    private static Pattern classFileNamePattern, classFullNameInPackagePattern, digitsClassnamePattern, sublimeJavaClassPattern;
    static
    {
        classFileNamePattern = Pattern.compile(CLASS_FILE_NAME_RE);
        classFullNameInPackagePattern = Pattern.compile(CLASS_FULL_NAME_IN_PKG_RE);
        digitsClassnamePattern = Pattern.compile(DIGITS_CLASS_NAME_RE);
        sublimeJavaClassPattern = Pattern.compile(SUBLIME_JAVA_CLASS_RE);
    }

    private static String getImport(String classFileName, String searchClass)
    {
        Matcher classnameMatcher = classFileNamePattern.matcher(classFileName);
        if (classnameMatcher.find())
        {
            String classname = classnameMatcher.group(1);
            if ((searchClass != null && searchClass.equals(classname)) ||
                (searchClass == null && !digitsClassnamePattern.matcher(classname).matches()))
            {
                String fullClassname = classFileName.replace("\\","/").replace("/", ".").replace(".class", "");
                if (fullClassname.startsWith("."))
                {
                    fullClassname = fullClassname.substring(1);
                }
                // Only allow class as import if it has a package. Classes in
                // default package are not allowed to be imported.
                if (classFullNameInPackagePattern.matcher(fullClassname).matches()){
                    return fullClassname;
                }
            }
        }
        return null;
    }

    private static String[] getClasspathEntries()
    {
        Set<String> paths = new HashSet<String>();
        paths.add("java/lang/String.class");
        for (String s : System.getProperty("java.class.path").split(System.getProperty("path.separator")))
        {
            paths.add(s);
        }
        return paths.toArray(new String[0]);
    }

    private static URL getUrlFromClasspathEntry(ClassLoader classLoader, String classpathEntry, String packagePath)
    {
        URL url = null;
        if (classpathEntry.endsWith(".class"))
            url = classLoader.getResource(classpathEntry);
        else
        {
            String path = "file://" + new File(classpathEntry).getAbsolutePath();
            if (path.endsWith(".jar"))
            {
                path = "jar:" + path + "!/" + packagePath;
            }
            else
            {
                path += "/" + packagePath;
            }
            try
            {
                System.err.println("path: " + path);
                url = new URL(path);
            }
            catch (Exception e)
            {}
        }
        return url;
    }

    private static void printImportsNotInJar(File current, File root, String classname, Set<String> possibleImports)
    {
        Matcher classnameMatcher = classFileNamePattern.matcher(current.getName());
        if (current.isFile() && classnameMatcher.matches())
        {
            if (classname == null || classname.equals(classnameMatcher.group(1)))
            {
                String classFileName = current.getAbsolutePath().substring(root.getAbsolutePath().length());
                printPossibleImport(getImport(classFileName, classname), possibleImports);
            }
        }
        else if (current.isDirectory())
        {
            for (File file : current.listFiles())
            {
                printImportsNotInJar(file, root, classname, possibleImports);
            }
        }
    }

    private static void getPossibleImports(String classname)
        throws IOException
    {
        Set<String> possibleImports = new HashSet<String>();
        ClassLoader classLoader = Thread.currentThread().getContextClassLoader();
        for (String s : getClasspathEntries())
        {
            URL url = getUrlFromClasspathEntry(classLoader, s, "");
            if (url == null)
                continue;
            System.err.println("s: " + s);
            System.err.println("url: " + url);

            String filename = URLDecoder.decode(url.getFile(), "UTF-8");
            if (url.getProtocol().equals("jar"))
            {
                filename = filename.substring(5, filename.indexOf("!"));

                JarFile jf = new JarFile(filename);
                Enumeration<JarEntry> entries = jf.entries();
                while (entries.hasMoreElements())
                {
                    String imp = getImport(entries.nextElement().getName(), classname);
                    printPossibleImport(imp, possibleImports);

                }
            }
            else
            {
                File folder = new File(filename);
                printImportsNotInJar(folder, folder, classname, possibleImports);
            }
        }
    }

    private static void printPossibleImport(String fullClassname, Set<String> possibleImports)
    {
        if (fullClassname != null && possibleImports.add(fullClassname))
            System.out.println(fullClassname);
    }

    private static void completePackage(String packageName)
        throws IOException
    {
        String packagePath = packageName.replace(".", "/");

        ClassLoader classLoader = Thread.currentThread().getContextClassLoader();
        for (String s : getClasspathEntries())
        {
            URL url = getUrlFromClasspathEntry(classLoader, s, packagePath);
            if (url == null)
                continue;
            System.err.println("s: " + s);
            System.err.println("packagename: " + packagePath);
            System.err.println("url: " + url);

            String filename = URLDecoder.decode(url.getFile(), "UTF-8");
            Set<String> packages = new HashSet<String>();
            if (url.getProtocol().equals("jar"))
            {
                filename = filename.substring(5, filename.indexOf("!"));

                JarFile jf = new JarFile(filename);
                Enumeration<JarEntry> entries = jf.entries();
                while (entries.hasMoreElements())
                {
                    String name = entries.nextElement().getName();
                    if (name.startsWith(packagePath))
                    {
                        name = name.substring(packagePath.length()+1);
                        int idx = name.indexOf('/');
                        if (idx != -1)
                        {
                            name = name.substring(0, idx);
                            if (packages.add(name))
                            {
                                System.out.println(name + "\tpackage" + sep + name);
                            }
                            continue;
                        }
                        name = name.replace(".class", "").replace('/', '.');
                        System.out.println(name + "\tclass" + sep + name);
                    }
                }
            }
            else
            {
                File folder = new File(filename);
                File[] files = folder.listFiles();
                if (files == null)
                    continue;
                for (File f : files)
                {
                    String name = f.getName();
                    if (name.endsWith(".class"))
                    {
                        name = name.substring(0, name.length()-6);
                        System.out.println(name + "\tclass" + sep + name);
                    }
                    else if (f.isDirectory())
                    {
                        System.out.println(name + "\tpackage" + sep + name);
                    }
                }
            }
        }
    }

    private static <T extends Throwable> void reportError(T e)
    {
        String type = "Error";
        if (e instanceof Exception)
            type = "Exception";
        String cn = e.getClass().getName();
        String msg = e.getMessage();
        if (msg == null)
            msg = "";

        StringWriter sw = new StringWriter();
        PrintWriter pw = new PrintWriter(sw);
        e.printStackTrace(pw);

        String stack = sw.toString();
        if (stack.indexOf(cn) == -1 && msg.indexOf(cn) == -1)
        {
            stack = cn + "\n" + stack;
        }
        if (stack.indexOf(msg) != -1)
            msg = "";
        System.err.println(type + ": " + msg + " " + stack + "\n" + ";;--;;");
    }

    private static class MyClassLoader
    extends ClassLoader
    {
        MyClassLoader()
        {
            super(MyClassLoader.class.getClassLoader());
        }
        public Class<?> loadClass(String name)
            throws ClassNotFoundException
        {
            // First we use the superclass to get a hold of the class.
            Class c = super.loadClass(name);
            DataInputStream s = null;
            try
            {
                // However, as the default class loader does not reload classes that have changed
                // we use the handle to the class to try and get it's actual .class definition
                // which we then define in THIS ClassLoader subclass. Hence: dynamic class reloading :)
                String path = ClassLoader.getSystemResource(c.getName().replaceAll("\\.", "/") + ".class").getFile();
                File f = new File(path);
                if (!f.exists()) {
                    return c;
                }
                int len = (int) f.length();
                byte[] data = new byte[len];
                s = new DataInputStream(new FileInputStream(f));
                s.readFully(data);
                return defineClass(name, data, 0, len);
            }
            catch (Exception e)
            {
                reportError(e);
            }
            finally
            {
                if (s != null)
                    try { s.close(); } catch (Exception e) {}
            }
            throw new ClassNotFoundException();
        }
    }

    private static Class<?> loadClass(String name)
        throws ClassNotFoundException
    {
        // See http://stackoverflow.com/questions/4285855/difference-betweeen-loading-a-class-using-classloader-and-class-forname/7099453#7099453
        // for an explanation of using a class loader vs Class.forName
        ClassLoader cls = new MyClassLoader();
        return cls.loadClass(name);
    }

    public static void main(String... unusedargs)
    {
        try
        {
            BufferedReader in = new BufferedReader(new InputStreamReader(System.in));
            boolean first = true;
            while (true)
            {
                try
                {
                    if (!first)
                        // Just to indicate that there's no more output from the command and we're ready for new input
                        System.out.println(";;--;;");
                    first = false;
                    String cmd = in.readLine();
                    if (cmd == null)
                        break;
                    String args[] = cmd.split(";;--;;");
                    System.err.println(args.length);
                    for (int i = 0; i < args.length; i++)
                    {
                        System.err.println(args[i]);
                    }

                    try
                    {
                        if (args[0].equals("-quit"))
                        {
                            System.err.println("quitting upon request");
                            return;
                        }
                        else if (args[0].equals("-possibleimports"))
                        {
                            String arg = null;
                            if (args.length > 1)
                            {
                                arg = args[1];
                            }
                            getPossibleImports(arg);
                            continue;
                        }
                        else if (args[0].equals("-findclass"))
                        {
                            String line = null;
                            ArrayList<String> packages = new ArrayList<String>();
                            try
                            {
                                while ((line = in.readLine()) != null)
                                {
                                    if (line.compareTo(sep) == 0)
                                        break;
                                    packages.add(line);
                                }
                            }
                            catch (Exception e)
                            {
                            }
                            boolean found = false;
                            for (String pack : packages)
                            {
                                try
                                {
                                    String classname = getClassname(pack, args[1]);
                                    System.err.println("Testing for: " + classname);
                                    Class c = loadClass(classname);
                                    System.out.println("" + c.getName());
                                    found = true;
                                    break;
                                }
                                catch (Exception e)
                                {
                                }
                            }
                            if (found)
                                continue;
                            // Still haven't found anything, so try to see if it's an internal class
                            for (String pack : packages)
                            {
                                String classname = getClassname(pack, args[1]);
                                while (!found && classname != null && classname.indexOf('.') != -1)
                                {
                                    int idx = classname.lastIndexOf('.');
                                    classname = classname.substring(0, idx) + "$" + classname.substring(idx+1);
                                    try
                                    {
                                        System.err.println("Testing for: " + classname);
                                        Class c = loadClass(classname);
                                        System.out.println("" + c.getName());
                                        found = true;
                                        break;
                                    }
                                    catch (Exception e)
                                    {
                                    }
                                }
                                if (found)
                                    break;
                            }
                            // Nothing yet.. Is it a package by any chance?
                            if (isPackage(args[1]))
                                System.out.println(args[1]);
                            continue;
                        }
                        if (args.length < 2)
                            continue;
                        Class<?> c = loadClass(args[1]);
                        String filter = "";
                        if (args.length >= 3)
                        {
                            filter = args[2];
                            if (filter.equals(sep))
                            {
                                filter = "";
                            }
                        }
                        int len = args.length - 3;
                        if (len < 0)
                            len = 0;
                        String[] templateParam = new String[len];
                        for (int i = 0; i < len; i++)
                        {
                            templateParam[i] = args[i+3];
                        }
                        if (args[0].equals("-complete"))
                        {
                            dumpCompletions(c.getFields(), filter, templateParam);
                            dumpCompletions(c.getDeclaredFields(), filter, templateParam);
                            dumpCompletions(c.getMethods(), filter, templateParam);
                            dumpCompletions(c.getDeclaredMethods(), filter, templateParam);
                            dumpCompletions(c.getClasses(), filter, templateParam);
                            dumpCompletions(c.getDeclaredClasses(), filter, templateParam);
                        }
                        else if (args[0].equals("-returntype"))
                        {
                            if (getReturnType(c.getDeclaredFields(), filter, templateParam))
                                continue;
                            if (getReturnType(c.getDeclaredMethods(), filter, templateParam))
                                continue;
                            if (getReturnType(c.getDeclaredClasses(), filter, templateParam))
                                continue;
                            if (getReturnType(c.getFields(), filter, templateParam))
                                continue;
                            if (getReturnType(c.getMethods(), filter, templateParam))
                                continue;
                            if (getReturnType(c.getClasses(), filter, templateParam))
                                continue;
                        }
                    }
                    catch (ClassNotFoundException x)
                    {
                        // Maybe it's a package then?
                        if (args[0].equals("-complete"))
                        {
                            completePackage(args[1]);
                        }
                        else if (args[0].equals("-returntype"))
                        {
                            String name = args[1] + "." + args[2];
                            if (isPackage(name))
                                System.out.println(name);
                        }
                    }
                }
                catch (Error e)
                {
                    reportError(e);
                }
                catch (Exception e)
                {
                    reportError(e);
                }
            }
        }
        catch (Exception e)
        {
            reportError(e);
        }
    }
}
